### ISSUE TITLE
Send optimistic transactionThreadReportID and createdReportActionIDForThread with API commands RequestMoney, CreateDistanceRequest, CompleteSplitBill, SendMoney

### ISSUE THREAD

**Issue by cead22** [Expensify/App#31411]
- [Design Doc](https://docs.google.com/document/d/1iGvpCoQ6Wt-i2wcpvaMKsWZRNs-oAsWoaS_nLue7_E0/edit)
- The back end changes will be done in a separate issue
- Update getMoneyRequestInformation to generate an optimistic transactionThreadReportID and createdReportActionIDForThread using `ReportUtils.buildTransactionThread`
   - Here's a [commit](https://github.com/Expensify/App/commit/481e3ec123fb66fb87b0a668cde8fc55aa0c2f69) with a proof-of-concept for these changes
       - We should make sure we're updating optimisticData with both values correctly
       - Ignore the comment `// TODO add logic to create transaction thread for each DM between split participants`
- Update getSendMoneyParams like getMoneyRequestInformation, send transactionThreadReportID and createdReportActionIDForThread, and add this to the request's optimisticData
- For the case of CompleteSplitBill where the user enters details manually before SmartScan finishes, we’ll generate random reportIDs in the client
    - Create a transactionThread and a created report action optimistically before the call to `buildOnyxDataForMoneyRequest` inside `completeSplitBill` like we do in `getMoneyRequestInformation`
    - Pass `optimisticCreatedAction.reportID` as the last param to the `buildOnyxDataForMoneyRequest` call
    - Add `transactionThreadReportID: transactionThread.reportID` and `createdReportActionIDForThread: optimisticCreatedAction.reportActionID` to the params object we pass to `API.write(‘CompleteSplitBill’, {params})`

**Comment by cead22** [comment]
@cdanwards can you comment here so I can assign you?

**Comment by cdanwards** [comment]
Commenting for Assignment

**Comment by cdanwards** [comment]
No update.

**Comment by cdanwards** [comment]
Next in line for work

**Comment by cdanwards** [comment]
Beginning work on this

**Comment by cdanwards** [comment]
Getting ready for PR

**Comment by paultsimura** [comment]
@cead22 @cdanwards can we assume these changes will fix the following issue?
A scan request is placed with an optimistic `transactionThreadReportID`, but when the `RequestMoney` API response comes, this `transactionThreadReportID` is replaced with another (which comes from BE). As a result, if the optimistic transaction thread report was opened, it becomes invalid and greyed out:

https://github.com/Expensify/App/assets/12595293/8aff586b-5991-4209-96e7-51c1734e9ad4

**Comment by cdanwards** [comment]
@cead22 I've been having some issues with the tests for IOU while implementing this. I made an assumption that we'd also need to pass both `optimisticCreatedActionReportID` and `optimisticCreatedActionfor to `buildOnyxDataForMoneyRequest`. Like so:
```
function buildOnyxDataForMoneyRequest(
   ...
   transactionThread,
    transactionThreadReportAction,
) {
  const optimisticData = [
  ...
        {
            onyxMethod: Onyx.METHOD.SET,
            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${transactionThread.reportID}`,
            value: {
                [transactionThreadReportAction.reportActionID]: transactionThreadReportAction,
            },
        },
  ...
```

And then update the necessary success and failure data.

So a few questions:
1) Is this the correct way to implement this?
2) If so, is there a good way to isolate the created transaction thread for testing [like we do here](https://github.com/Expensify/App/blob/21f5875d476159c26bf0dd83b12794c2c1f2944b/tests/actions/IOUTest.js#L548) for iou and chat reports?
3) If not, what report should the created action be attributed to?

**Comment by cead22** [comment]
Started a thread on slack to discuss in real time https://expensify.slack.com/archives/C01GTK53T8Q/p1702070081420789

**Comment by cdanwards** [comment]
Was prioritizing some other issues but this is next to finish

**Comment by lindboe** [comment]
Working on this now

**Comment by lindboe** [comment]
@cead22 Do you want to change the assignee to me?

Working on this today, will be based on the PR for #31092

**Comment by lindboe** [comment]
@cead22 After going through the example proof of concept and Dan's initial stab at this, I have some questions because it's still ambiguous to me how data should be passed to Onyx, with regards to two topics: 1. failure/success data, based on trying to extrapolate from similar existing updates, and 2. conditional behavior about whether a report is "new" does not make sense to me and isn't clear if we need it here.

## Failure and success data

Looking at some of the similar updates in the onyx data, they have some behavior around setting error fields that I'm not sure whether we need for transaction threads or not. It's very hard to parse if this data would actually be used in the case of a transaction thread, or if it has transaction-thread-specific-data, and @cead22 I'm wondering if you know? Examples:
1. if `optimisticTransactionThread` is analogous to the other reports in `buildOnyxDataForMoneyRequest`, it seems like it should:
    1. [set `errorFields` if it fails to be created](https://github.com/Expensify/App/blob/main/src/libs/actions/IOU.js#L463-L481) (note: I don't know why IOU report is resetting pending fields but chat report isn't, there) 
    2. [reset `errorFields` on success](https://github.com/Expensify/App/blob/main/src/libs/actions/IOU.js#L390-L392)
    3. It is ambiguous to me if the optimistic data should be setting `pendingFields` for the transaction thread like chat report and iou report do
2. If `optimisticCreatedActionForTransactionThread` is analogous to the other created report actions, it should [set `errors` in failure data](https://github.com/Expensify/App/blob/main/src/libs/actions/IOU.js#L509), and reset `errors` and `pendingAction` in success data
3. It would be easy enough to copy what other reports/report actions are doing in `buildOnyxDataForMoneyRequest`, but they're also handled differently in `getSendMoneyParams`:
    1. ~~~the optimistic data for reports does not set `pendingFields`~~ (it actually does if it's a new chat, but in a different way, and this was hard to spot... same for a few other differences I thought I spotted. This code is difficult to parse.)
4. `CompleteSplitBill` has a `REPORT_ACTIONS` update to add failure data for the chat report, but no corresponding success data that could ever reset that error, which added to my suspicion that this may not actually matter?

## isNewChat

In `buildOnyxDataForMoneyRequest`, there are the params `isNewChatReport` and `isNewIOUReport`, which is used to decide how to update onyx data. `getSendMoneyParams` also has a similar internal parameter (`isNewChat`). Do we need to have a similar parameter for "is new transaction thread", or can we assume it's always new? (As far as I've seen it's always new, but I would've assumed the same for `chatReport` and `iouReport`).

**Comment by lindboe** [comment]
Picking this back up today

**Comment by lindboe** [comment]
We're putting this on hold until #31448 is merged, since that introduces functionality needed to test this meaningfully

**Comment by cdanwards** [comment]
On hold until #31448 is merged

**Comment by lindboe** [comment]
Still on hold.

**Comment by cdanwards** [comment]
Working on this now.

**Comment by lindboe** [comment]
@cead22 @cdanwards and I had a question about this part: 
> Add transactionThreadReportID: transactionThread.reportID and createdReportActionIDForThread: optimisticCreatedAction.reportActionID to the params object we pass to API.write(‘CompleteSplitBill’, {params})

We're creating multiple transaction threads for splits in a loop, but we only send one `params` object. It looks like for things we create multiples of on this command, there's a `splits` key. Is that where we should be adding the IDs? So it would look like:

```
API.write(
  'CompleteSplitBill,
  {
    ...
    splits:
    [{
      transactionThreadReportID: ...,
      createdReportActionIDForThread: ...,
     },
     {
     ...
     }],
  });
```

**Comment by cdanwards** [comment]
Working on finalizing and getting up a PR!

**Comment by cead22** [comment]
@johncschuster please issue payment to C+ @situchan for [PR](https://github.com/Expensify/App/pull/35449) review. Thanks!

**Comment by johncschuster** [comment]
@situchan I've extended an invite, [here](https://www.upwork.com/ab/applicants/1762575110672728064/job-details). Can you accept that? Thanks!

**Comment by johncschuster** [comment]
@cead22 it looks like this issue was linked to a deploy blocker. Has that been resolved, or do I need to wait on issuing payment?

**Comment by paultsimura** [comment]
> @cead22 it looks like this issue was linked to a deploy blocker. Has that been resolved, or do I need to wait on issuing payment?

The PR review of this issue was blocked by that deploy blocker. Now that the blocker was fixed, we'll get back to testing the PR.

**Comment by situchan** [comment]
This issue was mentioned [here](https://github.com/Expensify/App/issues/37296#issuecomment-1966674360) which made Melvin post above message. Not regression.

**Comment by johncschuster** [comment]
Payment Summary:

Contributor: @cdanwards - Contractor
C+ Reviewer: @situchan - $500 - [Upwork job offer](https://www.upwork.com/nx/wm/offer/101283167/)

**Comment by johncschuster** [comment]
Paid!

**Comment by BiswaJK** [comment]
Triggered auto assignment to @cead22 , see https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=BugZero%20process%20steps%20for%20feature%20requests for more details.
